scene/gui/color_picker.cpp

```diff
@@ -1015,20 +1215,59 @@ void ColorPicker::_w_input(const Ref<InputEvent> &p_event) {
 	}
 }
 
-void ColorPicker::_preset_input(const Ref<InputEvent> &p_event, const Color &p_color) {
+void ColorPicker::_slider_or_spin_input(const Ref<InputEvent> &p_event) {
+	if (line_edit_mouse_release) {
+		line_edit_mouse_release = false;
+		return;
+	}
+	Ref<InputEventMouseButton> bev = p_event;
+	if (bev.is_valid() && !bev->is_pressed() && bev->get_button_index() == MouseButton::LEFT) {
+		_add_recent_preset_button(true);
+	}
+}
+
+void ColorPicker::_line_edit_input(const Ref<InputEvent> &p_event) {
+	Ref<InputEventMouseButton> bev = p_event;
+	if (bev.is_valid() && !bev->is_pressed() && bev->get_button_index() == MouseButton::LEFT) {
+		line_edit_mouse_release = true;
+	}
+}
+
+void ColorPicker::_preset_input(const Ref<InputEvent> &p_event, ColorPresetButton *p_preset) {
 	Ref<InputEventMouseButton> bev = p_event;
 
 	if (bev.is_valid()) {
 		if (bev->is_pressed() && bev->get_button_index() == MouseButton::LEFT) {
-			set_pick_color(p_color);
-			emit_signal(SNAME("color_changed"), p_color);
+			set_pick_color(p_preset->get_preset_color());
+			_add_recent_preset_button(false);
+			if (selected_preset) {
+				selected_preset->set_pressed(false);
+			}
```

Comment:
You could assign a ButtonGroup to the preset buttons to avoid manually managing the presses (also a group can give you the currently pressed button).